server/data/data-jmap/src/main/java/org/apache/james/jmap/api/filtering/Rule.java

```diff
@@ -268,45 +268,33 @@ public String toString() {
 
     public static class Builder {
 
-        private Optional<Id> id;
-        private Optional<String> name;
-        private Optional<Condition> condition;
-        private Optional<Action> action;
-
-        public Builder() {
-            this.id = Optional.empty();
-            this.name = Optional.empty();
-            this.condition = Optional.empty();
-            this.action = Optional.empty();
-        }
+        private Id id;
+        private String name;
+        private Condition condition;
+        private Action action;
 
         public Builder id(Id id) {
-            this.id = Optional.of(id);
+            this.id = id;
             return this;
         }
 
         public Builder name(String name) {
-            this.name = Optional.of(name);
+            this.name = name;
             return this;
         }
 
         public Builder condition(Condition condition) {
-            this.condition = Optional.of(condition);
+            this.condition = condition;
             return this;
         }
 
         public Builder action(Action action) {
-            this.action = Optional.of(action);
+            this.action = action;
             return this;
         }
 
         public Rule build() {
-            Preconditions.checkState(id.isPresent(), "`id` is mandatory");
-            Preconditions.checkState(name.isPresent(), "`name` is mandatory");
-            Preconditions.checkState(condition.isPresent(), "`condition` is mandatory");
-            Preconditions.checkState(action.isPresent(), "`action` is mandatory");
-
-            return new Rule(id.get(), name.get(), condition.get(), action.get());
+            return new Rule(id, name, condition, action);
```

Comment:
If everything is mandatory, why not doing a staged builder then instead, which should force it even more ?

Reply:
In my point of view, if simple builder pattern can handle it why i need more than that?
You can educate me what the benefit of `staged builder` for this case, pls.

Reply:
Is the builder still usefull now that all the validations are done in the constructor and all the field are mandatory ?

Reply:
Well a normal builder, you can forget a field, it will raise an exception, as all your fields are mandatory. A staged builder, you will be forced to go through all the steps of the builder, initializing all your mandatory fields, before being able to build. It's safer to avoid exceptions

Reply:
> Is the builder still usefull now that all the validations are done in the constructor and all the field are mandatory ?

Yes it still useful.
- All fields still mandatory - no matter where to put the validation.
- This Builder used in 2 part: business code(once and full step right after parsing json) and test.

My point:
- We provided: public constructor without validating ( mean fields not mandatory) but builder with validation (all fields mandatory) =>>> that make no sense to me.
- If we make private constructor + validation at builder =>>> making separate field rule off `business model`.

Reply:
> Well a normal builder, you can forget a field, it will raise an exception, as all your fields are mandatory
- With stage builder you still can pass `null/empty` value which invalid >>> you force nothing here.And in the test, you need to flexible to build object with missing thing to make sure everything handled well.
- In this case, `JmapRuleDTO` object is parse directly from json (body of JmapRequest) so i can even initial `Rule` object directly from constructor with business's validation rule.
- The builder most used from test

Reply:
you can still check for null and empty values in your build. but if it's only used in tests then i agree it's not that important then :)